Source/JavaScriptCore/runtime/LiteralParser.cpp

```diff
@@ -1273,8 +1300,35 @@ JSValue LiteralParser<CharType>::parseRecursively(VM& vm, uint8_t* stackLimit)
             if (std::optional<uint32_t> index = parseIndex(ident)) {
                 object->putDirectIndex(m_globalObject, index.value(), value);
                 RETURN_IF_EXCEPTION(scope, { });
-            } else
-                object->putDirect(vm, ident, value);
+            } else {
+                // When creating JSON object in this fast path, we know the following.
+                //   1. The object is definitely JSFinalObject.
+                //   2. The object rarely has duplicate properties.
+                //   3. Many same-shaped objects would be created from JSON. Thus very likely, there is already an existing Structure.
+                // Let's make the above case super fast, and fallback to the normal implementation when it is not true.
+                auto* structure = object->structure();
+                PropertyOffset offset = 0;
+                Structure* newStructure = nullptr;
+                if (LIKELY(!structure->isDictionary() && (newStructure = Structure::addPropertyTransitionToExistingStructure(structure, ident, 0, offset)))) {
+                    Butterfly* newButterfly = object->butterfly();
+                    if (structure->outOfLineCapacity() != newStructure->outOfLineCapacity()) {
+                        ASSERT(newStructure != structure);
+                        newButterfly = object->allocateMoreOutOfLineStorage(vm, structure->outOfLineCapacity(), newStructure->outOfLineCapacity());
+                        object->nukeStructureAndSetButterfly(vm, structure->id(), newButterfly);
+                    }
+
+                    validateOffset(offset);
+                    ASSERT(newStructure->isValidOffset(offset));
+
+                    // This assertion verifies that the concurrent GC won't read garbage if the concurrentGC
+                    // is running at the same time we put without transitioning.
+                    ASSERT(!object->getDirect(offset) || !JSValue::encode(object->getDirect(offset)));
+                    object->putDirectOffset(vm, offset, value);
+                    object->setStructure(vm, newStructure);
+                    ASSERT(!newStructure->mayBePrototype()); // There is no way to make it prototype object.
```

Comment:
Is this true? What happens if you do:  ```
let o1 = JSON.parse("{ a: 1 }");
o2.__proto__ = o1;
JSON.parse("{ a: 1 }");```

Reply:
Yes, this is true, since `mayBePrototype` bit gets moved to Structure, and setting `__proto__` causes structure transition now.